File: arkgrid/vision/matcher.py
"""Core template matching wrappers.

Ported from the TypeScript project's ``matcher.ts`` – wraps
``cv2.matchTemplate`` with ROI support, threshold filtering, and
best-of-N matching across a dictionary of templates.
"""

from dataclasses import dataclass
from typing import Dict, Optional, Tuple

import cv2
import numpy as np

from .constants import Roi


@dataclass
class MatchResult:
    """Result of a single template match."""
    key: str
    score: float
    location: Tuple[int, int]          # (x, y) top-left in full frame coords
    template_size: Tuple[int, int]     # (w, h)


def find_template(
    frame: np.ndarray,
    template: np.ndarray,
    roi: Optional[Roi] = None,
    method: int = cv2.TM_CCOEFF_NORMED,
) -> Tuple[float, Tuple[int, int]]:
    """Find *template* in *frame* (optionally within *roi*).

    Returns ``(score, (x, y))`` where the location is in full-frame
    coordinates (ROI offset is added back).
    """
    ox, oy = 0, 0
    target = frame
    if roi is not None:
        rx, ry, rw, rh = roi
        # Clamp to frame bounds
        rx = max(0, rx)
        ry = max(0, ry)
        rw = min(rw, frame.shape[1] - rx)
        rh = min(rh, frame.shape[0] - ry)
        if rw <= 0 or rh <= 0:
            return 0.0, (0, 0)
        target = frame[ry:ry + rh, rx:rx + rw]
        ox, oy = rx, ry

    # Template must fit inside target
    if template.shape[0] > target.shape[0] or template.shape[1] > target.shape[1]:
        return 0.0, (0, 0)

    result = cv2.matchTemplate(target, template, method)
    _, max_val, _, max_loc = cv2.minMaxLoc(result)

    # Convert location back to full-frame coords
    x = max_loc[0] + ox
    y = max_loc[1] + oy
    return float(max_val), (x, y)


def find_best_match(
    frame: np.ndarray,
    templates: Dict[str, np.ndarray],
    roi: Optional[Roi] = None,
    threshold: float = 0.8,
    method: int = cv2.TM_CCOEFF_NORMED,
) -> Optional[MatchResult]:
    """Try all *templates*, return the best scoring match above *threshold*.

    This mirrors the TypeScript ``getBestMatch()`` function from
    ``matcher.ts``, which iterates atlas entries and returns the highest
    scoring match.
    """
    best: Optional[MatchResult] = None

    for key, tmpl in templates.items():
        score, loc = find_template(frame, tmpl, roi=roi, method=method)
        if score >= threshold and (best is None or score > best.score):
            best = MatchResult(
                key=key,
                score=score,
                location=loc,
                template_size=(tmpl.shape[1], tmpl.shape[0]),
            )

    return best


def find_all_matches(
    frame: np.ndarray,
    templates: Dict[str, np.ndarray],
    roi: Optional[Roi] = None,
    threshold: float = 0.8,
    method: int = cv2.TM_CCOEFF_NORMED,
) -> list[MatchResult]:
    """Return *all* template matches above *threshold*, sorted by score descending."""
    matches: list[MatchResult] = []

    for key, tmpl in templates.items():
        score, loc = find_template(frame, tmpl, roi=roi, method=method)
        if score >= threshold:
            matches.append(MatchResult(
                key=key,
                score=score,
                location=loc,
                template_size=(tmpl.shape[1], tmpl.shape[0]),
            ))

    matches.sort(key=lambda m: m.score, reverse=True)
    return matches

File: web/src/lib/cv/matcher.ts
/**
 * Template matching wrappers — port of arkgrid/vision/matcher.py.
 *
 * Uses TM_CCOEFF_NORMED + minMaxLoc (same method as the Python source).
 * All intermediate Mats are deleted to avoid WASM memory leaks.
 * Environment-agnostic: no node:fs, only opencv.js via getCv().
 */

import { getCv } from './cvRuntime';
import type { Roi } from './constants';

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface MatchResult {
  key: string;
  score: number;
  loc: { x: number; y: number };
}

// ---------------------------------------------------------------------------
// findTemplate
// ---------------------------------------------------------------------------

/**
 * Find `template` in `frame`, optionally restricted to `roi` (x, y, w, h).
 *
 * Returns `{ score, x, y }` where x/y are in full-frame coordinates (the ROI
 * offset is added back, mirroring Python's `find_template`).
 *
 * Returns score 0 if the template is larger than the (possibly clipped) target,
 * or if the clamped ROI is empty.
 */
export function findTemplate(
  frame: any,
  template: any,
  roi?: Roi,
): { score: number; x: number; y: number } {
  const cv = getCv();

  let ox = 0;
  let oy = 0;
  let target: any = frame;
  let roiMat: any = null;

  if (roi !== undefined) {
    // Clamp to frame bounds (mirrors Python: max(0,rx), min(rw, cols-rx), etc.)
    let [rx, ry, rw, rh] = roi;
    rx = Math.max(0, rx);
    ry = Math.max(0, ry);
    rw = Math.min(rw, frame.cols - rx);
    rh = Math.min(rh, frame.rows - ry);

    if (rw <= 0 || rh <= 0) {
      return { score: 0, x: 0, y: 0 };
    }

    roiMat = frame.roi(new cv.Rect(rx, ry, rw, rh));
    target = roiMat;
    ox = rx;
    oy = ry;
  }

  try {
    // Template must fit inside target
    if (template.rows > target.rows || template.cols > target.cols) {
      return { score: 0, x: 0, y: 0 };
    }

    const result = new cv.Mat();
    try {
      cv.matchTemplate(target, template, result, cv.TM_CCOEFF_NORMED);
      const { maxVal, maxLoc } = cv.minMaxLoc(result);
      return {
        score: maxVal,
        x: maxLoc.x + ox,
        y: maxLoc.y + oy,
      };
    } finally {
      result.delete();
    }
  } finally {
    if (roiMat !== null) {
      roiMat.delete();
    }
  }
}

// ---------------------------------------------------------------------------
// findBestMatch
// ---------------------------------------------------------------------------

/**
 * Try all templates in `templates`, return the highest-scoring one ≥ threshold.
 * Returns null if no template clears the threshold.
 *
 * Mirrors Python's `find_best_match`.
 */
export function findBestMatch(
  frame: any,
  templates: Map<string, any>,
  roi?: Roi,
  threshold = 0.8,
): MatchResult | null {
  let best: MatchResult | null = null;

  for (const [key, tmpl] of templates) {
    const { score, x, y } = findTemplate(frame, tmpl, roi);
    if (score >= threshold && (best === null || score > best.score)) {
      best = { key, score, loc: { x, y } };
    }
  }

  return best;
}
